File: shared/src/sim/projectiles.ts
// Server-authoritative projectile sim + explosion resolution (PRD §24.3).
// Swept-sphere stepping makes tunneling impossible regardless of speed. Pure:
// the server steps these each tick; the client renders cosmetic predicted copies.

import type { Vec3 } from '../math';
import {
  v3,
  set,
  copy,
  addScaled,
  scale,
  reflect,
  normalize,
  clamp,
  lerpScalar,
  horizontalLength,
  rayCapsule,
  EPSILON,
} from '../math';
import { TUNING } from '../tuning';
import type { TraceWorld, EntityId } from './traceworld';

// Swept-sphere radius for projectile-vs-world contact (PRD §24.3). Small so the
// rocket/grenade reads as a point but never tunnels through thin geometry.
const PROJ_RADIUS = 0.15;

// ---- module-level scratch (zero-alloc hot paths) ----
const _delta = v3(); // vel * dt this step
const _toPlayer = v3(); // player.center - center
const _dir = v3(); // normalized blast direction

export type ProjKind = 'rocket' | 'grenade';

export interface Projectile {
  id: number;
  kind: ProjKind;
  pos: Vec3;
  vel: Vec3;
  ownerId: EntityId;
  fuse: number; // seconds remaining (grenade); rockets detonate on contact
  alive: boolean;
}

/** Analytic player hitbox for explosion overlap (own-code, 2 players — no engine). */
export interface PlayerCapsule {
  id: EntityId;
  center: Vec3;
  radius: number;
  halfHeight: number; // cylinder half-height
}

export interface ExplosionHit {
  id: EntityId;
  damage: number;
  impulse: Vec3; // velocity delta to add to that player
}

export interface ProjectileStep {
  detonated: boolean;
  point: Vec3; // detonation/contact point (valid when detonated)
}

export function makeProjectile(
  id: number,
  kind: ProjKind,
  pos: Vec3,
  vel: Vec3,
  ownerId: EntityId,
  fuse: number,
): Projectile {
  return {
    id,
    kind,
    pos: copy(v3(), pos),
    vel: copy(v3(), vel),
    ownerId,
    fuse,
    alive: true,
  };
}

/**
 * Advance one projectile by dt. Rockets: detonate on first static contact.
 * Grenades: bounce (restitution) off static geometry, detonate when fuse <= 0.
 * Returns whether it detonated this step and where.
 *
 * IMPLEMENTED BY: projectiles TDD task.
 */
export function stepProjectile(
  p: Projectile,
  world: TraceWorld,
  dt: number,
  out: ProjectileStep,
): ProjectileStep {
  out.detonated = false;

  if (p.kind === 'rocket') {
    // Slight gravity so long rockets arc.
    p.vel.y -= TUNING.rocket.projGravity * dt;

    // Swept-sphere this step; first static contact detonates.
    scale(_delta, p.vel, dt);
    const hit = world.castSphere(p.pos, PROJ_RADIUS, _delta);
    if (hit) {
      copy(out.point, hit.point);
      copy(p.pos, hit.point);
      out.detonated = true;
      p.alive = false;
      return out;
    }

    addScaled(p.pos, p.pos, p.vel, dt);
    copy(out.point, p.pos);
    return out;
  }

  // Grenade: gravity, bounce off geometry (restitution), detonate on fuse.
  p.vel.y -= TUNING.grenade.projGravity * dt;

  scale(_delta, p.vel, dt);
  const hit = world.castSphere(p.pos, PROJ_RADIUS, _delta);
  if (hit) {
    // Move to the contact point, then reflect+damp the velocity off the surface.
    copy(p.pos, hit.point);
    reflect(p.vel, p.vel, hit.normal);
    scale(p.vel, p.vel, TUNING.grenade.restitution);
    // Kill residual horizontal jitter so a settled grenade sits still.
    if (horizontalLength(p.vel) < 0.5) {
      p.vel.x = 0;
      p.vel.z = 0;
    }
  } else {
    addScaled(p.pos, p.pos, p.vel, dt);
  }

  // Fuse burns regardless of contact; detonate in place when it expires.
  p.fuse -= dt;
  if (p.fuse <= 0) {
    copy(out.point, p.pos);
    out.detonated = true;
    p.alive = false;
    return out;
  }

  copy(out.point, p.pos);
  return out;
}

/**
 * Direct projectile-vs-player contact for the segment the projectile swept this
 * tick (`start` -> `end`). Returns the NEAREST non-owner player whose capsule the
 * swept sphere crosses, plus the contact point, or null. The owner is excluded so
 * a rocket never detonates on the muzzle it just left (rocket-jumps still work via
 * splash). Without this a rocket only detonates on static geometry and sails
 * straight THROUGH players in the open — the "shoot through the player" bug.
 */
export function projectilePlayerHit(
  start: Vec3,
  end: Vec3,
  players: PlayerCapsule[],
  ownerId: EntityId,
): { id: EntityId; point: Vec3 } | null {
  const dx = end.x - start.x;
  const dy = end.y - start.y;
  const dz = end.z - start.z;
  const segLen = Math.hypot(dx, dy, dz);
  if (segLen < EPSILON) return null;
  set(_dir, dx / segLen, dy / segLen, dz / segLen);

  let bestT = Infinity;
  let bestId: EntityId | null = null;
  for (const pl of players) {
    if (pl.id === ownerId) continue;
    // Capsule core base = center bottom; combined radius folds in the proj sphere.
    set(_toPlayer, pl.center.x, pl.center.y - pl.halfHeight, pl.center.z);
    const t = rayCapsule(start, _dir, _toPlayer, pl.halfHeight * 2, pl.radius + PROJ_RADIUS, segLen);
    if (t !== null && t < bestT) {
      bestT = t;
      bestId = pl.id;
    }
  }
  if (bestId === null) return null;
  return {
    id: bestId,
    point: { x: start.x + _dir.x * bestT, y: start.y + _dir.y * bestT, z: start.z + _dir.z * bestT },
  };
}

/**
 * Resolve an explosion: for each player in `players`, analytic sphere-vs-capsule
 * overlap -> damage falloff (linear) and the §4 knockback impulse. Owner takes
 * selfDamageScale damage but full selfKnockbackScale knockback. Damage requires
 * line-of-sight (world.raycast to center); knockback applies regardless (boost
 * tech stays forgiving). `kind` selects rocket vs grenade tuning.
 *
 * IMPLEMENTED BY: projectiles TDD task.
 */
export function computeExplosion(
  kind: ProjKind,
  center: Vec3,
  ownerId: EntityId,
  players: PlayerCapsule[],
  world: TraceWorld,
  directHitId?: EntityId,
): ExplosionHit[] {
  const tuning = kind === 'rocket' ? TUNING.rocket : TUNING.grenade;
  const radius = tuning.splashRadius;
  const out: ExplosionHit[] = [];

  for (const player of players) {
    // Nearest point on the player's vertical core segment to the blast center —
    // fairer than center-to-center (a near-foot blast on a tall capsule still
    // reads as close). Clamp center.y to [segBottom, segTop].
    const segBottom = player.center.y - player.halfHeight;
    const segTop = player.center.y + player.halfHeight;
    const nearestY = center.y < segBottom ? segBottom : center.y > segTop ? segTop : center.y;
    // Vector from blast center to the player's nearest core point; its length is
    // the fairness distance used for falloff and the LOS raycast.
    set(_toPlayer, player.center.x - center.x, nearestY - center.y, player.center.z - center.z);
    const dist = Math.hypot(_toPlayer.x, _toPlayer.y, _toPlayer.z);
    if (dist > radius) continue;

    // Linear falloff: full damage at center, min at the radius edge.
    const t = clamp(dist / radius, 0, 1);
    let damage = lerpScalar(tuning.splashDamageMax, tuning.splashDamageMin, t);
    if (kind === 'rocket' && directHitId !== undefined && player.id === directHitId) {
      damage += TUNING.rocket.directDamage;
    }

    // Blast direction toward the player's nearest core point; straight up if the
    // blast is dead-center (degenerate).
    const len = normalize(_dir, _toPlayer);
    if (len < EPSILON) set(_dir, 0, 1, 0);

    // Line-of-sight gate on damage only: a wall between the blast and the player
    // nulls damage but knockback survives (boost tech stays forgiving, §24.3).
    const blocked = len >= EPSILON && world.raycast(center, _dir, dist) !== null;
    if (blocked) damage = 0;

    let impulseMag = tuning.knockback * (1 - t);

    // Owner: reduced self-damage, but full (scaled) self-knockback regardless of LOS.
    if (player.id === ownerId) {
      damage *= tuning.selfDamageScale;
      impulseMag *= tuning.selfKnockbackScale;
    }

    const rounded = Math.round(damage);
    if (rounded <= 0 && impulseMag < EPSILON) continue;

    const impulse = v3();
    scale(impulse, _dir, impulseMag);
    out.push({ id: player.id, damage: rounded, impulse });
  }

  return out;
}

File: godot/src/Shared/Sim/Projectiles.cs
// Server-authoritative projectile sim + explosion resolution. Swept-sphere
// stepping makes tunneling impossible regardless of speed. Pure. Ported from
// shared/src/sim/projectiles.ts.

using System;
using System.Collections.Generic;
using static WebRivals.Shared.VMath;

namespace WebRivals.Shared.Sim;

public enum ProjKind { Rocket, Grenade }

public sealed class Projectile
{
    public int Id;
    public ProjKind Kind;
    public Vec3 Pos = V3();
    public Vec3 Vel = V3();
    public int OwnerId;
    public double Fuse;   // seconds remaining (grenade); rockets detonate on contact
    public bool Alive;
}

/// <summary>Analytic player hitbox for explosion overlap.</summary>
public sealed class PlayerCapsule
{
    public int Id;
    public Vec3 Center;
    public double Radius;
    public double HalfHeight;
}

public sealed class ExplosionHit
{
    public int Id;
    public int Damage;
    public Vec3 Impulse;
}

public sealed class ProjectileStep
{
    public bool Detonated;
    public Vec3 Point = V3();
}

public sealed class PlayerHitResult
{
    public int Id;
    public Vec3 Point;
}

public static class Projectiles
{
    private const double PROJ_RADIUS = 0.15;

    private static readonly Vec3 _delta = V3();
    private static readonly Vec3 _toPlayer = V3();
    private static readonly Vec3 _dir = V3();

    public static Projectile MakeProjectile(int id, ProjKind kind, Vec3 pos, Vec3 vel, int ownerId, double fuse)
    {
        return new Projectile
        {
            Id = id,
            Kind = kind,
            Pos = Copy(V3(), pos),
            Vel = Copy(V3(), vel),
            OwnerId = ownerId,
            Fuse = fuse,
            Alive = true,
        };
    }

    /// <summary>Advance one projectile by dt. Rockets detonate on first static
    /// contact; grenades bounce off geometry and detonate when fuse expires.</summary>
    public static ProjectileStep StepProjectile(Projectile p, ITraceWorld world, double dt, ProjectileStep outStep)
    {
        outStep.Detonated = false;

        if (p.Kind == ProjKind.Rocket)
        {
            p.Vel.Y -= Tuning.Rocket.ProjGravity * dt;

            Scale(_delta, p.Vel, dt);
            var hit = world.CastSphere(p.Pos, PROJ_RADIUS, _delta);
            if (hit != null)
            {
                Copy(outStep.Point, hit.Point);
                Copy(p.Pos, hit.Point);
                outStep.Detonated = true;
                p.Alive = false;
                return outStep;
            }

            AddScaled(p.Pos, p.Pos, p.Vel, dt);
            Copy(outStep.Point, p.Pos);
            return outStep;
        }

        // Grenade
        p.Vel.Y -= Tuning.Grenade.ProjGravity * dt;

        Scale(_delta, p.Vel, dt);
        var ghit = world.CastSphere(p.Pos, PROJ_RADIUS, _delta);
        if (ghit != null)
        {
            Copy(p.Pos, ghit.Point);
            Reflect(p.Vel, p.Vel, ghit.Normal);
            Scale(p.Vel, p.Vel, Tuning.Grenade.Restitution);
            if (HorizontalLength(p.Vel) < 0.5)
            {
                p.Vel.X = 0;
                p.Vel.Z = 0;
            }
        }
        else
        {
            AddScaled(p.Pos, p.Pos, p.Vel, dt);
        }

        p.Fuse -= dt;
        if (p.Fuse <= 0)
        {
            Copy(outStep.Point, p.Pos);
            outStep.Detonated = true;
            p.Alive = false;
            return outStep;
        }

        Copy(outStep.Point, p.Pos);
        return outStep;
    }

    /// <summary>Direct projectile-vs-player contact for the swept segment.
    /// Returns the NEAREST non-owner player whose capsule the swept sphere
    /// crosses, plus the contact point, or null.</summary>
    public static PlayerHitResult ProjectilePlayerHit(Vec3 start, Vec3 end, List<PlayerCapsule> players, int ownerId)
    {
        double dx = end.X - start.X;
        double dy = end.Y - start.Y;
        double dz = end.Z - start.Z;
        double segLen = Math.Sqrt(dx * dx + dy * dy + dz * dz);
        if (segLen < EPSILON) return null;
        Set(_dir, dx / segLen, dy / segLen, dz / segLen);

        double bestT = double.PositiveInfinity;
        int bestId = -1;
        bool found = false;
        foreach (var pl in players)
        {
            if (pl.Id == ownerId) continue;
            Set(_toPlayer, pl.Center.X, pl.Center.Y - pl.HalfHeight, pl.Center.Z);
            double? t = RayCapsule(start, _dir, _toPlayer, pl.HalfHeight * 2, pl.Radius + PROJ_RADIUS, segLen);
            if (t != null && t.Value < bestT)
            {
                bestT = t.Value;
                bestId = pl.Id;
                found = true;
            }
        }
        if (!found) return null;
        return new PlayerHitResult
        {
            Id = bestId,
            Point = new Vec3(start.X + _dir.X * bestT, start.Y + _dir.Y * bestT, start.Z + _dir.Z * bestT),
        };
    }

    /// <summary>Resolve an explosion: damage falloff (linear) + knockback impulse.
    /// Owner takes selfDamageScale damage but full selfKnockbackScale knockback.
    /// Damage requires line-of-sight; knockback applies regardless.</summary>
    public static List<ExplosionHit> ComputeExplosion(
        ProjKind kind, Vec3 center, int ownerId, List<PlayerCapsule> players, ITraceWorld world, int directHitId = -1)
    {
        double splashMax, splashMin, radius, knockback, selfDamageScale, selfKnockbackScale;
        if (kind == ProjKind.Rocket)
        {
            splashMax = Tuning.Rocket.SplashDamageMax;
            splashMin = Tuning.Rocket.SplashDamageMin;
            radius = Tuning.Rocket.SplashRadius;
            knockback = Tuning.Rocket.Knockback;
            selfDamageScale = Tuning.Rocket.SelfDamageScale;
            selfKnockbackScale = Tuning.Rocket.SelfKnockbackScale;
        }
        else
        {
            splashMax = Tuning.Grenade.SplashDamageMax;
            splashMin = Tuning.Grenade.SplashDamageMin;
            radius = Tuning.Grenade.SplashRadius;
            knockback = Tuning.Grenade.Knockback;
            selfDamageScale = Tuning.Grenade.SelfDamageScale;
            selfKnockbackScale = Tuning.Grenade.SelfKnockbackScale;
        }

        var outHits = new List<ExplosionHit>();

        foreach (var player in players)
        {
            double segBottom = player.Center.Y - player.HalfHeight;
            double segTop = player.Center.Y + player.HalfHeight;
            double nearestY = center.Y < segBottom ? segBottom : (center.Y > segTop ? segTop : center.Y);
            Set(_toPlayer, player.Center.X - center.X, nearestY - center.Y, player.Center.Z - center.Z);
            double dist = Math.Sqrt(_toPlayer.X * _toPlayer.X + _toPlayer.Y * _toPlayer.Y + _toPlayer.Z * _toPlayer.Z);
            if (dist > radius) continue;

            double t = Clamp(dist / radius, 0, 1);
            double damage = LerpScalar(splashMax, splashMin, t);
            if (kind == ProjKind.Rocket && directHitId != -1 && player.Id == directHitId)
                damage += Tuning.Rocket.DirectDamage;

            double len = Normalize(_dir, _toPlayer);
            if (len < EPSILON) Set(_dir, 0, 1, 0);

            bool blocked = len >= EPSILON && world.Raycast(center, _dir, dist) != null;
            if (blocked) damage = 0;

            double impulseMag = knockback * (1 - t);

            if (player.Id == ownerId)
            {
                damage *= selfDamageScale;
                impulseMag *= selfKnockbackScale;
            }

            int rounded = (int)Math.Floor(damage + 0.5); // JS Math.round semantics
            if (rounded <= 0 && impulseMag < EPSILON) continue;

            var impulse = V3();
            Scale(impulse, _dir, impulseMag);
            outHits.Add(new ExplosionHit { Id = player.Id, Damage = rounded, Impulse = impulse });
        }

        return outHits;
    }
}
